### ISSUE TITLE
raster masking gives unexpected output if nodata is nan

### ISSUE THREAD

**Issue by dcdenu4** [natcap/invest#738]
There have been a few user forum posts recently that have caused issues because the users input raster had a **nodata** value of `nan`. Should checking / warning of `nan` **nodata** values be a part of UI validation? Or a part of model validation? Or both?

The recent forum post where this came up: https://community.naturalcapitalproject.org/t/no-data-value-ndr-model/2345/7

**Comment by davemfish** [comment]
This sounds familiar, but I can't remember. Why is `nan` a problem?

**Comment by davemfish** [comment]
Related: https://github.com/natcap/invest/issues/629

**Comment by dcdenu4** [comment]
The forum user mentioned that they created the raster using R

**Comment by davemfish** [comment]
>  in this instance we have nan as the nodata value. With nan as nodata we get:

If it's properly set as the nodata value, then I would think our conventional nodata masking approach in raster calculator ops would avoid the "sum" or "mean" problems. Is it worth double-checking the masking code in the models where this came up?

**Comment by emlys** [comment]
```
>>> numpy.isclose(numpy.nan, numpy.nan)
False
```
this might complicate things 😕

**Comment by davemfish** [comment]
Aha!

**Comment by davemfish** [comment]
```
>>> numpy.isclose(numpy.nan, numpy.nan, equal_nan=True)
True
```

**Comment by dcdenu4** [comment]
>Is it worth double-checking the masking code in the models where this came up?

I think in the NDR case I was looking at you're right, `numpy.isclose(numpy.nan, numpy.nan, equal_nam=True)` would have solved the problem there.

**Comment by davemfish** [comment]
I think our discussion leads to treating this as a bug in mask operations that are ubiquitous in `raster_calculator` local ops. I added the milestone too because I think this is common enough to prioritize it. And it should be an easy fix by just adding the `equal_nan=True` arg to any `numpy.isclose` in those ops. 

If we also want regression tests to cover this, then this becomes a much larger effort. We probably don't have existing unit tests for many of those local operations.

**Comment by emlys** [comment]
#536 wouldn't solve the issue of missing unit tests, but would at least make it possible to run the existing tests with raster inputs that have a nodata value of `nan`

**Comment by davemfish** [comment]
> #536 wouldn't solve the issue of missing unit tests, but would at least make it possible to run the existing tests with raster inputs that have a nodata value of `nan`

That's a good idea. It does seem like that sort of property-based test pattern would work well for testing really small units like these local raster ops. 

The main question might be whether or not we want to hoist all those local_op functions to the top-level of the module so that they can be unit-tested.

**Comment by dcdenu4** [comment]
I can work on this assuming that we have settled on:

- adding the `equal_nan=True` arg to any `numpy.isclose` in `raster_calculator` like local ops.
- Punt on adding regression or unit tests for now

**Comment by phargogh** [comment]
Dang, the `nan` case is an interesting one.

I wonder if it's worth coming back to an idea that @emlys had some time ago: would it be worth having a `utils` function to compare nodata values?  We now have a couple interesting reasons why such a function might be worth it:

1. This `nan` case, requiring that we remember to pass an extra parameter to `numpy.isclose` every time we use it in the `raster_calculator` context and we have a user-defined nodata value
2. Integer value comparison (`a != b`) is noticeably faster in numpy than `numpy.isclose(a, b)`, but we can only do integer comparison on integer datasets and _sometimes_ users provide a floating-point LULC so it's a gamble to rely on integer equality.
3. It would make testing the function a whole lot more direct, while still leaving open the option to include `nan` comparison in hypothesis-based testing later on

A centralized implementation of this might be as simple as 
```python
def this_is_a_bad_name_but_lets_check_nodata(array, nodata):
    # comparing an integer array against numpy.nan works correctly
    if numpy.issubdtype(array.dtype, numpy.integer):
        return array != nodata
    return numpy.isclose(array, nodata, equal_nan=True)
```

We have one other example of a code snippet we use basically everywhere in `utils.make_suffix_string`, so maybe this would make sense for us?  I'm curious what everyone thinks.

**Comment by emlys** [comment]
I like that idea of having a `utils` function. It would be nice not to have to remember to check nodata in the right way every time, since there are now a couple steps.

**Comment by davemfish** [comment]
This definitely feels like the right solution to the testing problem.

**Comment by dcdenu4** [comment]
>This definitely feels like the right solution to the testing problem.

I agree. I'll take a crack at this implementation approach. Thanks for outlining and mocking that James.

**Comment by davemfish** [comment]
Fixed in #770

**Comment by dcdenu4** [comment]
Lately on the forum we've seen some issues around NoData values being `nan`. I'm re-opening this issue as it seems like the same problem. 

https://community.naturalcapitalproject.org/t/swy-error-on-task-calculate-curve-number-13/3291/4
https://community.naturalcapitalproject.org/t/ndr-model-fill-pits-creates-filled-dem-with-one-value/3342/6

**Comment by phargogh** [comment]
Related (at least to the pitfilling issue): https://github.com/natcap/pygeoprocessing/issues/248

**Comment by phargogh** [comment]
I think we can close this now that pitfilling (which now includes the `nan` nodata comparison fix) is in pygeoprocessing 2.4.0, which will be in the InVEST 3.13.0 release.
